tests/5.0/requires/test_requires_unified_shared_memory.c

```diff
@@ -0,0 +1,136 @@
+//===---test_requires_unified_shared_memory.c -------------------------------===//
+//
+// OpenMP API Version 5.0 Nov 2018
+//
+//
+////===----------------------------------------------------------------------===//
+#include <omp.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include "ompvv.h"
+
+#define N 1024
+
+#pragma omp requires unified_shared_memory
+
+int unified_shared_memory_scalar() {
+  OMPVV_INFOMSG("Unified shared memory testing - scalar value");
+  int errors = 0;
+  
+  int aVariable = 0;
+  // aVariable is only mapped to, with unified shared memory we should see the change reflected in the host.
+  // If no map is explicitely added, then scalars are firstprivate
+#pragma omp target map(to:aVariable)
+  {
+    aVariable += 10;
+  }
+
+  aVariable += 10;
+#pragma omp target map(alloc:aVariable)
+  {
+    aVariable += 10;
+  }
+
+  OMPVV_TEST_AND_SET_VERBOSE(errors, aVariable != 30);
+  return errors;
+}
+
+int unified_shared_memory_stack() {
+  OMPVV_INFOMSG("Unified shared memory testing - Array on stack");
+  int errors = 0;
+  
+  int anArray[N];
+  int anArrayCopy[N];
+
+  for (int i = 0; i < N; i++) {
+    anArray[i] = i;
+    anArrayCopy[i] = 0;
+  }
+  // If for any reason there is data mapping, we avoid the movement
+  // that may occur through the default mapping map(tofrom)
+#pragma omp target data map(alloc:anArray) 
+  {
+    // Modify again on the host
+    for (int i = 0; i < N; i++) {
+        anArray[i] += 10;
+    }
+#pragma omp target
+    {
+      for (int i = 0; i < N; i++) {
+        anArray[i] += 10;
+      }
+    }
+  }
+  // Modify again on the host
+  for (int i = 0; i < N; i++) {
+    anArray[i] += 10;
+  }
+#pragma omp target map(alloc:anArray)
+  {
+    for (int i = 0; i < N; i++) {
+      anArrayCopy[i] = anArray[i];
+    }
+  }
+  for (int i = 0; i < N; i++) {
+    OMPVV_TEST_AND_SET_VERBOSE(errors, anArray[i] != i + 30);
+    OMPVV_TEST_AND_SET_VERBOSE(errors, anArrayCopy[i] != i + 30);
+    if (errors) break;
+  }
+  return errors;
+}
+
+int unified_shared_memory_heap() {
+  OMPVV_INFOMSG("Unified shared memory testing - Array on heap");
+  int errors = 0;
+  
+  int *anArray;
+  int *anArrayCopy;
+
+  anArray = (int*)malloc(sizeof(int)*N);
+  anArrayCopy = (int*)malloc(sizeof(int)*N);
+
+  for (int i = 0; i < N; i++) {
+    anArray[i] = i;
+    anArrayCopy[i] = 0;
+  }
+  // Modiify in the device
+#pragma omp target
+  {
+    for (int i = 0; i < N; i++) {
+      anArray[i] += 10;
+    }
+  }
+  // Modify again on the host
+  for (int i = 0; i < N; i++) {
+    anArray[i] += 10;
+  }
+
+  // Get the value the device is seeing
+#pragma omp target 
+  {
+    for (int i = 0; i < N; i++) {
+      anArrayCopy[i] = anArray[i];
+    }
+  }
+
+  for (int i = 0; i < N; i++) {
+    OMPVV_TEST_AND_SET_VERBOSE(errors, anArray[i] != i + 30);
+    OMPVV_TEST_AND_SET_VERBOSE(errors, anArrayCopy[i] != i + 30);
```

Comment:
Why 30? Don't you only add 10 twice?

Reply:
Ah! Yes, copy and paste mistake

Reply:
Does this test case, the heap one, fail as well with clang?

Reply:
Yes, It fails with:
Libomptarget fatal error 1: failure of target construct while offloading is mandatory

Reply:
This is summit 9.0.0. 

I was under the impression that this was same trunk, but this version output makes me doubt that. 
```
 clang --version
clang version 9.0.0 (https://github.com/clang-ykt/clang 5e1ea7eba87adbe43fed5746bcbba41d7cff1bf2) (https://github.com/clang-ykt/llvm fa0da850cf08eaffb84cb1480a27efa63c63fd14)
Target: powerpc64le-unknown-linux-gnu
Thread model: posix
InstalledDir: /sw/summit/llvm/9.0.0/9.0.0.patch001.cuda-10.1.168/bin
```

Reply:
Can we try trunk?

Reply:
I would need to re-compile it on our UD system for sm_50 or higher I guess. It only seems to have sm_30, and it fails with 
```
Target architecture does not support unified addressing
```

Reply:
Uhmm, on a second thought, we have K40's. They support managed memory, but they don't seem to have a page migration engine. i am not sure if you are relying on this for llvm ?